tools/workload-simulator.html

```diff
@@ -0,0 +1,1160 @@
+      }
+
+      // Upload data using queue.writeBuffer.
+      if (bufferSubData.value > 0) {
+        if (!uploadBuffer || uploadBufferSize != bufferSubDataArray.byteLength) {
+          if (uploadBuffer) uploadBuffer.destroy();
+          uploadBufferSize = bufferSubDataArray.byteLength;
+          uploadBuffer = device.createBuffer({
+            size: bufferSubDataArray.byteLength,
+            usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
+          });
+        }
+        device.queue.writeBuffer(uploadBuffer, 0, bufferSubDataArray);
+      }
+
+      // Upload data using mapAsync and a queue of staging buffers.
+      if (mapAsyncArray.length > 0) {
+        if (mapAsyncBufferSize != mapAsyncArray.byteLength) {
+          mapAsyncBufferSize = mapAsyncArray.byteLength;
+          if (mapAsyncBuffer) mapAsyncBuffer.destroy();
+          mapAsyncBuffer = device.createBuffer({size: mapAsyncArray.byteLength, usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST});
+        }
+        let buffer = null;
+        while (mapAsyncReady.length) {
+          buffer = mapAsyncReady.shift();
+          if (buffer.size == mapAsyncArray.byteLength)
+            break;
+          buffer.destroy();
+          mapAsyncBuffersOutstanding--;
+          buffer = null;
+          bufferWarning.textContent = '';
+        }
+        if (!buffer) {
+          buffer = device.createBuffer({
+            size: mapAsyncArray.byteLength,
+            usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE,
+            mappedAtCreation: true,
+          });
+          buffer.size = mapAsyncArray.byteLength;
+          mapAsyncBuffersOutstanding++;
+          if (mapAsyncBuffersOutstanding > 10) {
+            bufferWarning.textContent = ` Warning: mapAsync requests from ${mapAsyncBuffersOutstanding} frames ago have not resolved yet. ${(mapAsync.value * mapAsyncBuffersOutstanding).toFixed(2)} MB of staging buffers allocated.`;
+          }
+        }
+        new Float32Array(buffer.getMappedRange()).set(mapAsyncArray);
+        buffer.unmap();
+        const commandEncoder = device.createCommandEncoder();
+        commandEncoder.copyBufferToBuffer(buffer, 0, mapAsyncBuffer, 0, mapAsyncArray.byteLength);
+        // TODO: combine this submit with the main one, but we'll have to delay calling mapAsync until after the submit.
+        device.queue.submit([commandEncoder.finish()]);
+        // TODO: use this data during rendering.
+        buffer.mapAsync(GPUMapMode.WRITE).then(()=>mapAsyncReady.push(buffer));
+      }
+
+      // Record command buffer.
+      const commandEncoder = device.createCommandEncoder();
+      let renderAttachment = swapChainView;
+      if (multisampling.checked) {
+        renderAttachment = multisampleRenderAttachment.createView();
+      }
+      const passEncoder = commandEncoder.beginRenderPass({
+        colorAttachments: [ {
+          attachment: renderAttachment,
```

Comment:
I think Chrome has fixed it, so no need to use `attachment` any more

Reply:
Yes, I just noticed this today. I'll remove it.